[LimKaiWei] iP

LimKaiWei opened the pull request:
# Elliot
___Elliot___ frees your mind of having to remember things you need to do. It's,
> _Bing Chilling_
- text-based 📖 
- easy to learn 💭 
- ~~FAST~~ SUPER FAST to use 🚗 
- All you need to do is, 🎃 

download it from [here](https://github.com/LimKaiWei/ip/releases).
1. double-click it.
2. add your tasks.
3. let it manage your tasks for you 😉
4. And it is __FREE__!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

- --
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

LimKaiWei committed:
Level-0 commit
- changed from Duke to Elliot

LimKaiWei committed:
Level-1 commit
- updated gitignore to ignore .swp for vim, hopefully works after this push

LimKaiWei committed:
Implement Level-2

LimKaiWei committed:
Implement Level-3
- import ImList (immutable list class) from cs2030

LimKaiWei committed:
Implement Level-4 with more types of tasks

LimKaiWei committed:
Added test cases
- updated Elliot with `stripStringArray` function to get rid of the whitespaces when splitting strings
- updated Elliot to handle no description after command

LimKaiWei committed:
Update conditions for Level-5 error handling
- handle no description error
- handle unidentified commands/tasks outside of current scope

LimKaiWei committed:
Implemented delete task function and included more test

LimKaiWei committed:
clean up code

LimKaiWei committed:
Implement Level-7 saving task list to file

LimKaiWei committed:
Merge branch 'branch-Level-7'

LimKaiWei committed:
Implement Level-8 string to date object
- bug from eventTask unable to differentiate /from and /to dates, bot assumes first date is /from and second date /to when placed /to in front of /from

LimKaiWei committed:
Merge branch 'branch-Level-8'

LimKaiWei committed:
Overhaul structure of code
simultaneously moved into package level, use enum, improve oop

LimKaiWei committed:
Merge branch 'branch-A-MoreOOP'

LimKaiWei committed:
Merge branch 'master' into add-gradle-support

LimKaiWei committed:
Edit gradle build file

LimKaiWei committed:
Merge branch 'add-gradle-support'

LimKaiWei committed:
Add some JUnit test cases

LimKaiWei committed:
Merge branch 'branch-A-JUnit'

LimKaiWei committed:
Build fat jar

Tkaixiang commented on src/main/java/utility/Ui.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    public static void introSay() {
+        String logo = "    ________    __    ________  ______\n"
+                + "   / ____/ /   / /   /  _/ __ \\/_  __/\n"
+                + "  / __/ / /   / /    / // / / / / /\n"
+                + " / /___/ /___/ /____/ // /_/ / / /\n"
+                + "/_____/_____/_____/___/\\____/ /_/\n";
+        String intro = "Hello! I'm\n" + logo
```
Very nice touch :D

Tkaixiang commented on src/main/java/elliot/Elliot.java:
```diff
@@ -0,0 +1,68 @@
+            String[] commandString = stripStrArray(userInput.toLowerCase().split(" ", 2));
+            Ui.say("");
+            try {
+                CommandType commandType = CommandType.parseStringToCommand(commandString[0]);
+                Command command = commandType.getCommand();
+                command = command.parseArguments(commandString.length < 2 
+                        ? "" 
```
Cool way of handling command types using enums, but perhaps the parsing of arguments could use Regex for greater flexibility :D?

LimKaiWei replied:
good idea! will experiment with regex if possible in subsequent

Tkaixiang commented on src/main/java/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,52 @@
+            Ui.say("which number task would you like me to mark undone?\n");
+            return taskList;
+        } 
+        for (int i : listToUnmark) {
+            if (taskList.isValidIndex(i - 1)) {
+                taskList = taskList.markTaskAsUndone(i - 1);
+                Ui.say("OK, I've marked this task as not done yet:\n"
+                        + taskList.get(i - 1).toString() + "\n");
+            } else {
+                Ui.say("index out of range\n");
+            }
```
Could break out of the loop when a valid task has already been found

Tkaixiang commented on src/main/java/utility/TaskList.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.List;
+import java.io.Serializable;
+
+public class TaskList implements Serializable {
+    private final ImList<Task> imTaskList;
+
+    public TaskList() {
+        imTaskList = new ImList<Task>();
+    }
```
Interesting implementation!

Tkaixiang reviewed (approved):
LGTM, nice code :)

yanqiyqh commented on src/main/java/utility/ImList.java:
```diff
@@ -0,0 +1,253 @@
+ * @author  cs2030
+ * @param <E> the type of elements in this list
+ */
+public class ImList<E> implements Iterable<E>, Serializable {
```
Didn't expect to see familiar stuff from 2030 here :O

LimKaiWei replied:
honestly idk if allowed to use i probably should ask first...

yanqiyqh commented on src/main/java/elliot/Elliot.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private String[] stripStrArray(String[] strArray) {
+        return Arrays.stream(strArray)
```
I like the use of `stream` here, respectable.

yanqiyqh commented on src/main/java/task/DeadlineTask.java:
```diff
@@ -0,0 +1,39 @@
+    @Override
+    public Task markAsUndone() {
+        return super.isDone
+            ? new DeadlineTask(false, super.taskDescription, this.deadlineDateTime)
```
I like this formatting, very concise compared to if-else statements

yanqiyqh reviewed (commented):
Overall, I don't spot any code quality violations. Great work!

LimKaiWei committed:
Add comments to all classes

LimKaiWei committed:
Merge branch 'branch-A-JavaDoc'

LimKaiWei committed:
clean up elliot class
- remove unused imports and functions

LimKaiWei committed:
Build javadocs via gradle
- fix some errors in header comments
- get gradle configured for javadoc task

LimKaiWei committed:
Merge branch 'branch-A-JavaDoc'

LimKaiWei committed:
Tweak code to comply with coding standard

LimKaiWei committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/command/Command.java
#	src/main/java/utility/Ui.java

LimKaiWei committed:
Implement Level-9 find command

LimKaiWei committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/task/Task.java
#	src/main/java/utility/TaskList.java

LimKaiWei committed:
Import and configure checkstyle

LimKaiWei committed:
Adhere to checkstyle

LimKaiWei committed:
Merge branch 'A-CheckStyle'

LimKaiWei committed:
Implement GUI
up till right before enhancements

LimKaiWei committed:
Fully integrate elliot to GUI
noted bugs
intro text (fixed using mono)
first user input invalid command (fixed using default commandType var as LIST type)

todo/bugs
bye command will exit before printing dialog box
test findcommand

LimKaiWei committed:
Fix checkstyle

LimKaiWei committed:
Merge branch 'Level-10'

LimKaiWei committed:
Modified visuals for GUI

No image background.

To make the GUI slightly more visually appealing.

Let's add a background image for the main window and adjust the size
accordingly.

LimKaiWei committed:
Add assertions

No assertions made anywhere in code base.

This commit introduces assertions to validate different variables
throughout the code. The assertions help catch potential issues early
and improve overall code reliability.

Let's add assertions for:
- deadline command to have none empty task descriptions and argument
  length of 2
- delete/mark/unmark command to have none empty lists and that the index
  is valid
- elliot to have empty string upon flushing ui
- Storage to have valid save file

LimKaiWei committed:
Follow coding standard

No coding standard.

Since we have learnt something about coding standard lets apply it.

Let's
- remove empty catch blocks from commandtype, deadline, event, find,
  todo commands by refactoring each ui.say and Elliot
- rearrange code in findcommand for guard clause and happy path
- event, deadline, todo command remove recycle variable

LimKaiWei committed:
Implement Streams for collating tasks strings

Minimal streams used.

Want to include more streams.

Let's use streams to collect individual tasks strings for task list
toString method.

LimKaiWei committed:
Merge pull request #1 from LimKaiWei/branch-A-Assertions

Add assertions to improve code readability

LimKaiWei committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/command/DeadlineCommand.java

LimKaiWei committed:
Merge pull request #2 from LimKaiWei/branch-A-CodeQuality

Follow coding standard

LimKaiWei committed:
Merge pull request #3 from LimKaiWei/branch-A-Streams

Implement Streams for collating tasks strings

LimKaiWei committed:
Setup Github actions

No Github actions setup yet.

Github actions can automate workflows like testing when pushing commits.

Let's add workflow for gradle.

LimKaiWei committed:
Integrate tagging to tasks via hashtags

The code does not have any tagging capabilities.

Let's implement tagging and tagging search.

LimKaiWei committed:
Merge branch 'branch-C-Tagging'

LimKaiWei committed:
Acknowledge reuse of code from CS2030

specifically ImList.java

LimKaiWei committed:
Set title to window for GUI stage scene as "Elliot"

LimKaiWei committed:
Update and improve user guide markdown for elliot chatbot

LimKaiWei committed:
Include how to use section in the user guide

LimKaiWei committed:
Fix user guide markdown github pages not correctly displayed

LimKaiWei committed:
Fix user guide markdown github pages not correctly displayed

LimKaiWei committed:
Fix correct output display for 'find *' command #7

- Fixed unintended behavior when using the 'find *' command.
- Ensured the command now displays the correct output.
- Added tests to verify the correct functionality of the 'find *'
  command.

Closes #7

LimKaiWei committed:
Merge pull request #8 from LimKaiWei/7-bug-find-command-does-not-display-output-correctly

Fix correct output display for 'find *' command #7

LimKaiWei committed:
Fix correct TaskList display issue when list is empty

- Fixed bug where TaskList does not display properly when the list
  command is executed from a fresh start or after all tasks are deleted.
- Ensured TaskList shows an appropriate message when no tasks are
  present.

LimKaiWei committed:
Merge pull request #10 from LimKaiWei/9-bug-tasklist-not-displaying-properly

Fix correct TaskList display issue when list is empty
